pkg/cosign/tuf/client.go

```diff
@@ -0,0 +1,199 @@
+const TufLocalStore = ".sigstore/root/"
+const TufRemoteStore = "test-root-123"
+const TufRootKeys = ".sigstore/keys.json"
+const TufLocalTargets = TufLocalStore + "targets/"
+const TufThreshold = 3
+
+// Global TUF client. Stores local targets in .sigstore/root.
+// Could be in memory local store, but that would mean re-download each time cosign is run.
+var rootClient *client.Client
+var rootClientMu = &sync.Mutex{}
+
+type targetDestination struct {
+	*os.File
+}
+
+func (t *targetDestination) Delete() error {
+	t.Close()
+	return nil
+}
+
+type ByteDestination struct {
+	*bytes.Buffer
+}
+
+func (b *ByteDestination) Delete() error {
+	b.Reset()
+	return nil
+}
+
+func getRootKeys() ([]*data.Key, error) {
+	rootFile, err := os.Open(TufRootKeys)
+	if err != nil {
+		return nil, err
+	}
+	var rootKeys []*data.Key
+	if err := json.NewDecoder(rootFile).Decode(&rootKeys); err != nil {
+		return nil, err
+	}
+	return rootKeys, nil
+}
+
+// Gets the global TUF client if the directory exists.
+func RootClient(ctx context.Context) (*client.Client, error) {
+	rootClientMu.Lock()
+	defer rootClientMu.Unlock()
+	if rootClient == nil {
+		// Instantiate the global TUF client from the local store. This does not do a download.
+		local, err := tuf_leveldbstore.FileLocalStore(TufLocalStore)
+		if err != nil {
+			return nil, errors.Wrap(err, "initializing local store")
+		}
+		remote, err := GcsRemoteStore(ctx, TufRemoteStore, nil, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "initializing remote store")
+		}
+		rootClient = client.NewClient(local, remote)
+	}
+	return rootClient, nil
+}
+
+func updateMetadataAndDownloadTargets(c *client.Client) error {
+	// Download initial targets and store in .sigstore/root/targets/.
+	targetFiles, err := c.Update()
+	if err != nil && !client.IsLatestSnapshot(err) {
+		return errors.Wrap(err, "updating tuf metadata")
+	}
+
+	// Download targets, if they don't already exist and match the updated metadata.
+	for name := range targetFiles {
+		if err := downloadTarget(name, c, nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func downloadTarget(name string, c *client.Client, out client.Destination) error {
+	f, err := os.Create(TufLocalTargets + name)
+	if err != nil {
+		return errors.Wrap(err, "creating target file")
+	}
+	defer f.Close()
+	dest := targetDestination{f}
+
+	if err := c.Download(name, &dest); err != nil {
+		return errors.Wrap(err, "downloading target")
+	}
+	if out != nil {
+		_, err = io.Copy(out, dest)
+	}
+	return err
+}
+
+// Instantiates the global TUF client. Downloads all initial targets and stores in .sigstore/root/targets/.
+func Init(ctx context.Context) error {
+	rootClient, err := RootClient(ctx)
+	if err != nil {
+		return errors.Wrap(err, "initializing root client")
+	}
+	rootKeys, err := getRootKeys()
+	if err != nil {
+		return errors.Wrap(err, "retrieving root keys")
+	}
+	if err := rootClient.Init(rootKeys, TufThreshold); err != nil {
```

Comment:
Pls make `TufThreshold` and `TufLocalTargets` args to init and/or provide a function that allows for non-default location and threshold

Reply:
Yep!! Exactly the WIP part, these are gonna be args to `cosign init` with those defaults

Reply:
the only problem one is the location of the default local store:
* if you run `cosign init --local path/to/some/other/directory`
* and then you run `cosign verify`: how is it supposed to carry over to the non-default path?

i'm not convinced you should be able to configure that part, unless it's pulled from an env var

Reply:
Pulling from an env var is pretty standard practice, e.g. `KUBECONFIG=` and `DOCKER_CONFIG=`